clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -3091,6 +3104,58 @@ static void handleMaxGlobalWorkDimAttr(Sema &S, Decl *D,
                                                                      E);
 }
 
+SYCLIntelLoopFuseAttr *
+Sema::mergeSYCLIntelLoopFuseAttr(Decl *D, const SYCLIntelLoopFuseAttr &Attr,
+                                 Expr *E) {
+  if (checkAttrMutualExclusion<SYCLIntelLoopFuseIndependentAttr>(*this, D,
+                                                                 Attr))
+    return nullptr;
+
+  if (D->hasAttr<SYCLIntelLoopFuseAttr>())
+    return nullptr;
+
+  return ::new (Context) SYCLIntelLoopFuseAttr(Context, Attr, E);
+}
+
+SYCLIntelLoopFuseIndependentAttr *Sema::mergeSYCLIntelLoopFuseIndependentAttr(
+    Decl *D, const SYCLIntelLoopFuseIndependentAttr &Attr, Expr *E) {
+  if (checkAttrMutualExclusion<SYCLIntelLoopFuseAttr>(*this, D, Attr))
+    return nullptr;
+
+  if (D->hasAttr<SYCLIntelLoopFuseIndependentAttr>())
+    return nullptr;
+
+  return ::new (Context) SYCLIntelLoopFuseIndependentAttr(Context, Attr, E);
+}
+
+// Handles loop_fuse and loop_fuse_independent.
+// These attributes are incompatible with eachother.
+template <typename AttrType, typename ConflictingAttrType>
+static void handleLoopFusionAttr(Sema &S, Decl *D, const ParsedAttr &Attr) {
+  if (D->isInvalidDecl())
+    return;
+
+  if (checkForDuplicateAttribute<AttrType>(S, D, Attr))
+    return;
+
+  if (checkAttrMutualExclusion<ConflictingAttrType>(S, D, Attr))
+    return;
+
+  unsigned NumArgs = Attr.getNumArgs();
+  if (NumArgs > 1) {
+    S.Diag(Attr.getLoc(), diag::warn_attribute_too_many_arguments) << Attr << 0;
+    return;
+  }
+
+  // Handle optional attribute argument.
+  if (Attr.isArgExpr(0))
+    // Attribute argument specified.
+    S.AddOneConstantValueAttr<AttrType>(D, Attr, Attr.getArgAsExpr(0));
+  else
+    // Attribute argument not specified.
+    D->addAttr(::new (S.Context) AttrType(S.Context, Attr));
```

Comment:
Because we can't use the default argument trick in Attr.td, we can still get the benefit by having the case where the expression isn't written in source create the default argument anyway. e.g., you can call `IntegerLiteral::Create()` to create an integer literal with the appropriate value in this case, and then further uses of the semantic attribute (like in CodeGen) can all assume that the attribute always has a valid expression argument.